File: apps/mobile/lib/services/coach/coach_narrative_service.dart
/// Coach Narrative Service — Sprint S35 / updated S44.
///
/// Generates 4 independent narrative components, each validated through
/// [ComplianceGuard] before reaching the UI:
///   1. Greeting — personalized, context-aware salutation
///   2. Score Summary — FRI score with trend explanation
///   3. Tip Narrative — actionable educational insight
///   4. Chiffre Choc Reframe — contextualizes shock figures
///
/// Synchronous API: uses [FallbackTemplates] as the generation engine
/// (no LLM dependency). If ComplianceGuard flags a violation, falls back
/// to the template output.
///
/// Async LLM-enhanced API (S44): delegates to [CoachOrchestrator] which
/// runs the full SLM → BYOK → FallbackTemplates chain.
///
/// All output text is in French (informal "tu"), free of banned terms,
/// and compliant with LSFin educational framing.
library;

import 'package:mint_mobile/services/coach_llm_service.dart';

import 'package:mint_mobile/l10n/app_localizations.dart';

import 'coach_models.dart';
import 'coach_orchestrator.dart';
import 'compliance_guard.dart';
import 'fallback_templates.dart';

/// Result container for all 4 narrative components.
class CoachNarrativeResult {
  /// Personalized greeting (max 30 words).
  final String greeting;

  /// FRI score summary with trend (max 80 words).
  final String scoreSummary;

  /// Actionable educational tip (max 120 words).
  final String tipNarrative;

  /// Chiffre choc contextualizer (max 100 words).
  final String chiffreChocReframe;

  const CoachNarrativeResult({
    required this.greeting,
    required this.scoreSummary,
    required this.tipNarrative,
    required this.chiffreChocReframe,
  });
}

/// Stateless service that generates compliant coach narratives.
///
/// Every generated text passes through [ComplianceGuard.validate()]
/// before reaching the UI. If validation fails (banned terms,
/// prescriptive language, hallucinated numbers, wrong language),
/// the fallback template is used instead.
class CoachNarrativeService {
  CoachNarrativeService._();

  // ═══════════════════════════════════════════════════════════════
  // Public API
  // ═══════════════════════════════════════════════════════════════

  /// Generate all 4 narrative components in one call.
  ///
  /// Each component is independently validated through ComplianceGuard.
  /// Failure of one component does not affect the others.
  static CoachNarrativeResult generateAll(CoachContext ctx, S s) {
    return CoachNarrativeResult(
      greeting: _generateGreeting(ctx, s),
      scoreSummary: _generateScoreSummary(ctx, s),
      tipNarrative: _generateTipNarrative(ctx, s),
      chiffreChocReframe: _generateChiffreChocReframe(ctx, s),
    );
  }

  /// Generate only the greeting component.
  static String generateGreeting(CoachContext ctx, S s) =>
      _generateGreeting(ctx, s);

  /// Generate only the score summary component.
  static String generateScoreSummary(CoachContext ctx, S s) =>
      _generateScoreSummary(ctx, s);

  /// Generate only the tip narrative component.
  static String generateTipNarrative(CoachContext ctx, S s) =>
      _generateTipNarrative(ctx, s);

  /// Generate only the chiffre choc reframe component.
  static String generateChiffreChocReframe(CoachContext ctx, S s) =>
      _generateChiffreChocReframe(ctx, s);

  /// Generate an enrichment guide for a specific data block.
  ///
  /// Returns a conversational prompt guiding the user to provide
  /// missing data for the given [blockType]. Uses the same
  /// fallback → compliance pipeline as other components.
  static String generateEnrichmentGuide(
    CoachContext ctx, {
    required String blockType,
    required S s,
  }) =>
      _generateEnrichmentGuide(ctx, blockType, s);

  // ═══════════════════════════════════════════════════════════════
  // LLM-enhanced API (S44) — delegates to CoachOrchestrator
  // ═══════════════════════════════════════════════════════════════

  /// Generate all 4 components with SLM → BYOK → template chain.
  ///
  /// Each component is independently generated through [CoachOrchestrator].
  /// [byokConfig] is optional; if null, only SLM and templates are tried.
  static Future<CoachNarrativeResult> generateAllEnhanced(
    CoachContext ctx, {
    LlmConfig? byokConfig,
    required S s,
  }) async {
    // Sequential calls: SlmEngine has an _isGenerating guard that blocks
    // concurrent SLM requests. Parallel Future.wait() would cause 3 of 4
    // components to fall back to templates even when SLM is available.
    final greeting = await CoachOrchestrator.generateNarrativeComponent(
      componentType: ComponentType.greeting,
      ctx: ctx,
      byokConfig: byokConfig,
      s: s,
    );
    final scoreSummary = await CoachOrchestrator.generateNarrativeComponent(
      componentType: ComponentType.scoreSummary,
      ctx: ctx,
      byokConfig: byokConfig,
      s: s,
    );
    final tip = await CoachOrchestrator.generateNarrativeComponent(
      componentType: ComponentType.tip,
      ctx: ctx,
      byokConfig: byokConfig,
      s: s,
    );
    final chiffreChoc = await CoachOrchestrator.generateNarrativeComponent(
      componentType: ComponentType.chiffreChoc,
      ctx: ctx,
      byokConfig: byokConfig,
      s: s,
    );
    return CoachNarrativeResult(
      greeting: greeting.text,
      scoreSummary: scoreSummary.text,
      tipNarrative: tip.text,
      chiffreChocReframe: chiffreChoc.text,
    );
  }

  // ═══════════════════════════════════════════════════════════════
  // Internal — generate + validate + fallback
  // ═══════════════════════════════════════════════════════════════

  static String _generateGreeting(CoachContext ctx, S s) {
    final text = FallbackTemplates.greeting(ctx, s);
    final result = ComplianceGuard.validate(
      text,
      context: ctx,
      componentType: ComponentType.greeting,
    );
    return result.useFallback ? FallbackTemplates.greeting(ctx, s) : result.sanitizedText;
  }

  static String _generateScoreSummary(CoachContext ctx, S s) {
    final text = FallbackTemplates.scoreSummary(ctx, s);
    final result = ComplianceGuard.validate(
      text,
      context: ctx,
      componentType: ComponentType.scoreSummary,
    );
    return result.useFallback
        ? FallbackTemplates.scoreSummary(ctx, s)
        : result.sanitizedText;
  }

  static String _generateTipNarrative(CoachContext ctx, S s) {
    final text = FallbackTemplates.tipNarrative(ctx, s);
    final result = ComplianceGuard.validate(
      text,
      context: ctx,
      componentType: ComponentType.tip,
    );
    return result.useFallback
        ? FallbackTemplates.tipNarrative(ctx, s)
        : result.sanitizedText;
  }

  static String _generateChiffreChocReframe(CoachContext ctx, S s) {
    final text = FallbackTemplates.chiffreChocReframe(ctx, s);
    final result = ComplianceGuard.validate(
      text,
      context: ctx,
      componentType: ComponentType.chiffreChoc,
    );
    return result.useFallback
        ? FallbackTemplates.chiffreChocReframe(ctx, s)
        : result.sanitizedText;
  }

  static String _generateEnrichmentGuide(CoachContext ctx, String blockType, S s) {
    final text = FallbackTemplates.enrichmentGuide(ctx, blockType, s);
    final result = ComplianceGuard.validate(
      text,
      context: ctx,
      componentType: ComponentType.enrichmentGuide,
    );
    return result.useFallback
        ? FallbackTemplates.enrichmentGuide(ctx, blockType, s)
        : result.sanitizedText;
  }
}

File: services/backend/app/services/coach/coach_narrative_service.py
"""
Coach Narrative Service — Sprint S35.

Generates 4 independent personalized narrative components:
    1. greeting — Dashboard welcome message
    2. score_summary — FRI score explanation
    3. tip_narrative — Actionable educational tip
    4. premier_eclairage_reframe — Confidence-anchored reframe

Architecture:
    - Currently uses FallbackTemplates (deterministic, no LLM).
    - Each component passes through ComplianceGuard validation.
    - When BYOK LLM is integrated, LLM output will be validated first;
      fallback templates are used when ComplianceGuard rejects LLM output.

Rules:
    - No banned terms (garanti, certain, optimal, etc.)
    - No prescriptive language (tu devrais, tu dois, il faut)
    - Educational tone, conditional language
    - Disclaimer on every response
    - All text in French (informal "tu")

Sources:
    - LSFin art. 3 (information financiere)
    - LPD art. 6 (protection des donnees)
    - FINMA circular 2008/21 (operational risk)
"""

from app.services.coach.coach_models import (
    CoachContext,
    CoachNarrativeResult,
    ComponentType,
)
from app.services.coach.compliance_guard import ComplianceGuard
from app.services.coach.fallback_templates import FallbackTemplates


STANDARD_DISCLAIMER = (
    "Outil educatif simplifie. "
    "Ne constitue pas un conseil financier (LSFin)."
)

STANDARD_SOURCES = [
    "LSFin art. 3 (information financiere)",
    "LPD art. 6 (protection des donnees)",
]


class CoachNarrativeService:
    """Generates personalized, compliance-validated coach narratives.

    Each generate_* method:
        1. Produces text via FallbackTemplates (deterministic).
        2. Validates through ComplianceGuard.
        3. Falls back to template if guard rejects (should not happen
           for deterministic templates, but safety net is in place).
    """

    def __init__(self):
        self._guard = ComplianceGuard()

    def generate_greeting(self, ctx: CoachContext) -> str:
        """Generate a personalized dashboard greeting.

        Args:
            ctx: CoachContext with user data.

        Returns:
            Compliant greeting string (max 30 words).
        """
        text = FallbackTemplates.greeting(ctx)
        result = self._guard.validate(
            text, context=ctx, component_type=ComponentType.greeting
        )
        if result.use_fallback:
            # Template should always be compliant; this is a safety net
            return FallbackTemplates.greeting(ctx)
        return result.sanitized_text

    def generate_score_summary(self, ctx: CoachContext) -> str:
        """Generate FRI score summary with trend.

        Args:
            ctx: CoachContext with FRI data.

        Returns:
            Compliant score summary string (max 80 words).
        """
        text = FallbackTemplates.score_summary(ctx)
        result = self._guard.validate(
            text, context=ctx, component_type=ComponentType.score_summary
        )
        if result.use_fallback:
            return FallbackTemplates.score_summary(ctx)
        return result.sanitized_text

    def generate_tip_narrative(self, ctx: CoachContext) -> str:
        """Generate an educational tip based on financial indicators.

        Args:
            ctx: CoachContext with financial indicators.

        Returns:
            Compliant tip string (max 120 words).
        """
        text = FallbackTemplates.tip_narrative(ctx)
        result = self._guard.validate(
            text, context=ctx, component_type=ComponentType.tip
        )
        if result.use_fallback:
            return FallbackTemplates.tip_narrative(ctx)
        return result.sanitized_text

    def generate_premier_eclairage_reframe(self, ctx: CoachContext) -> str:
        """Generate premier éclairage reframe anchored on confidence score.

        Args:
            ctx: CoachContext with confidence_score.

        Returns:
            Compliant reframe string (max 100 words).
        """
        text = FallbackTemplates.premier_eclairage_reframe(ctx)
        result = self._guard.validate(
            text, context=ctx, component_type=ComponentType.premier_eclairage
        )
        if result.use_fallback:
            return FallbackTemplates.premier_eclairage_reframe(ctx)
        return result.sanitized_text

    def generate_all(self, ctx: CoachContext) -> CoachNarrativeResult:
        """Generate all 4 components independently.

        Each component is generated and validated separately.
        Failure in one component does not affect others.

        Args:
            ctx: CoachContext with all user data.

        Returns:
            CoachNarrativeResult with all 4 components + metadata.
        """
        greeting = self.generate_greeting(ctx)
        score_summary = self.generate_score_summary(ctx)
        tip_narrative = self.generate_tip_narrative(ctx)
        premier_eclairage_reframe = self.generate_premier_eclairage_reframe(ctx)

        return CoachNarrativeResult(
            greeting=greeting,
            score_summary=score_summary,
            tip_narrative=tip_narrative,
            premier_eclairage_reframe=premier_eclairage_reframe,
            used_fallback={
                "greeting": True,
                "score_summary": True,
                "tip_narrative": True,
                "premier_eclairage_reframe": True,
            },
            disclaimer=STANDARD_DISCLAIMER,
            sources=list(STANDARD_SOURCES),
        )
